src/main/java/wtf/g4s8/examples/spaxos/Proposer.java

```diff
@@ -72,29 +75,42 @@
     /**
      * New proposer with server id and list of acceptors.
      */
-    public Proposer(final int server, final List<? extends Acceptor<T>> acc) {
-        this(Proposal.init(server), acc, new CompletableFuture<>());
+    public Proposer(final int server, final String transactionId, final List<? extends Acceptor<T>> acc) {
+        this(Proposal.init(server), transactionId, acc, new CompletableFuture<>(), false);
     }
 
-    private Proposer(final Proposal prop, final List<? extends Acceptor<T>> acceptors, final CompletableFuture<T> future) {
+    private Proposer(final Proposal prop, final String transactionId, final List<? extends Acceptor<T>> acceptors, final CompletableFuture<T> future, Boolean crashed) {
         this.prop = prop;
         this.acceptors = acceptors;
         this.future = future;
+        this.transactionId = transactionId;
+        this.crashed = crashed;
     }
 
     public Future<T> propose(final T value) {
+        if (crashed) {
+            return null;
+        }
         final Proposal next = this.prop.next();
-        Log.d("proposing %s : %s", next, value);
+        Log.d("[%s] proposing %s : %s", transactionId, next, value);
         final QuorumPrepared<T> callback = new QuorumPrepared<>(next, value, this.acceptors, this);
         this.acceptors.parallelStream().forEach(
                 acc -> acc.prepare(next, callback)
         );
-        EXEC_TIMEOUT.schedule(callback::timeout, 300 + RNG.nextInt(50), TimeUnit.MILLISECONDS);
+        EXEC_TIMEOUT.schedule(callback::timeout, Config.delay + RNG.nextInt(Config.delay/6), TimeUnit.MILLISECONDS);
         return this.future;
     }
 
     private Proposer<T> restart(Proposal prop) {
-        return new Proposer<>(prop, this.acceptors, this.future);
+        return new Proposer<>(prop, this.transactionId, this.acceptors, this.future, this.crashed);
+    }
+
+    public void kill() {
+        this.crashed = true;
+    }
+
+    public boolean isDead() {
```

Comment:
@stepanvalyavskiy I'm not sure about implementing proposer as a dedicated instance here - only the whole node could crash, not just a proposer. If we're implementing optimized version of Paxos-commit, then we have on one node: RM, proposer, N acceptors. This mean we have to kill all this services on node in case of crash.